src/mcpdotnet/Configuration/McpServerBuilderExtensions.Tools.cs

```diff
@@ -0,0 +1,249 @@
+            Type t when t.IsArray => "array",
+            Type t when t == typeof(DateTime) => "string",
+            _ => "object"
+        };
+    }
+
+    private static async Task<CallToolResponse> CallTool(RequestContext<CallToolRequestParams> request, MethodInfo method, CancellationToken cancellationToken)
+    {
+        var methodParameters = method.GetParameters();
+        List<object?> parameters = ResolveParameters(request, methodParameters, cancellationToken);
+
+        if (cancellationToken.IsCancellationRequested)
+            return new CallToolResponse { Content = [new Content { Text = "Operation was cancelled" }] };
+
+        try
+        {
+            using var scope = request.Server.ServiceProvider?.CreateScope();
+            var objectInstance = CreateObjectInstance(method, scope?.ServiceProvider);
+            var result = method.Invoke(objectInstance, parameters.ToArray());
+
+
+            if (result is Task task)
+            {
+                await task.ConfigureAwait(false);
+                var resultProperty = task.GetType().GetProperty("Result");
+                result = resultProperty?.GetValue(task);
+            }
+
+            if (result is string resultString)
+                return new CallToolResponse { Content = [new Content() { Text = resultString, Type = "text" }] };
+
+            if (result is string[] resultStringArray)
+                return new CallToolResponse { Content = resultStringArray.Select(s => new Content() { Text = s, Type = "text" }).ToList() };
+
+            if (result is null)
+                return new CallToolResponse { Content = [new Content() { Text = "null" }] };
+
+            if (result is JsonElement jsonElement)
+                return new CallToolResponse { Content = [new Content() { Text = jsonElement.GetRawText(), Type = "text" }] };
+
+            return new CallToolResponse { Content = [new Content() { Text = result.ToString(), Type = "text" }] };
+        }
+        catch (TargetInvocationException e)
+        {
+            throw new McpServerException(e.Message, e);
+        }
+    }
+
+    private static List<object?> ResolveParameters(RequestContext<CallToolRequestParams> request, ParameterInfo[] methodParameters, CancellationToken cancellationToken)
+    {
+        var parameters = new List<object?>(methodParameters.Length);
+
+        foreach (var parameter in methodParameters)
+        {
+            if (parameter.ParameterType == typeof(CancellationToken))
+            {
+                parameters.Add(cancellationToken);
+            }
+            else if (request.Params?.Arguments != null && request.Params.Arguments.TryGetValue(parameter.Name ?? "NoName", out var value))
+            {
+                if (value is JsonElement element)
+                    value = JsonSerializer.Deserialize(element.GetRawText(), parameter.ParameterType);
+
+                parameters.Add(Convert.ChangeType(value, parameter.ParameterType));
```

Comment:
Why is the ChangeType call needed if JsonSerializer.Deserialize returns successfully?

Reply:
Fixed in #54